src/vma/dev/ib_ctx_handler.cpp

```diff
@@ -166,49 +193,114 @@ ib_ctx_handler::~ib_ctx_handler() {
 		} ENDIF_VERBS_FAILURE;
 		m_umr_cq = NULL;
 	}
-	if (ibv_dealloc_pd(m_p_ibv_pd))
-		ibch_logdbg("pd deallocation failure (errno=%d %m)", errno);
+	if (m_p_ibv_pd) {
+		IF_VERBS_FAILURE(ibv_dealloc_pd(m_p_ibv_pd)) {
+			ibch_logdbg("pd deallocation failure (errno=%d %m)", errno);
+		} ENDIF_VERBS_FAILURE;
+		VALGRIND_MAKE_MEM_UNDEFINED(m_p_ibv_pd, sizeof(struct ibv_pd));
+		m_p_ibv_pd = NULL;
+	}
+
 	delete m_p_ctx_time_converter;
 	delete m_p_ibv_device_attr;
+
+	ibv_close_device(m_p_ibv_context);
+	m_p_ibv_context = NULL;
+
 	BULLSEYE_EXCLUDE_BLOCK_END
 }
 
+void ib_ctx_handler::set_str()
+{
+	char str_x[255] = {0};
+
+	m_str[0] = '\0';
+
+	str_x[0] = '\0';
+	sprintf(str_x, " %s:", get_ibname());
+	strcat(m_str, str_x);
+
+	str_x[0] = '\0';
+	sprintf(str_x, " port(s): %d", m_p_ibv_device_attr->phys_port_cnt);
+	strcat(m_str, str_x);
+
+	str_x[0] = '\0';
+	sprintf(str_x, " vendor: %d", m_p_ibv_device_attr->vendor_part_id);
+	strcat(m_str, str_x);
+
+	str_x[0] = '\0';
+	sprintf(str_x, " fw: %s", m_p_ibv_device_attr->fw_ver);
+	strcat(m_str, str_x);
+}
+
+void ib_ctx_handler::print_val()
+{
+	set_str();
+	ibch_logdbg("%s", m_str);
+}
+
 ts_conversion_mode_t ib_ctx_handler::get_ctx_time_converter_status()
 {
 	return m_p_ctx_time_converter->get_converter_status();
 }
 
-ibv_mr* ib_ctx_handler::mem_reg(void *addr, size_t length, uint64_t access)
+uint32_t ib_ctx_handler::mem_reg(void *addr, size_t length, uint64_t access)
 {
-	// Register the memory block with the HCA on this ibv_device
-	ibch_logfunc("(dev=%p) addr=%p, length=%d, m_p_ibv_pd=%p on dev=%p",
-			m_p_ibv_device, addr, length, m_p_ibv_pd,
-			m_p_ibv_pd->context->device);
+	struct ibv_mr *mr = NULL;
+	uint32_t lkey = (uint32_t)(-1);
+
 #ifdef DEFINED_IBV_EXP_ACCESS_ALLOCATE_MR
 	struct ibv_exp_reg_mr_in in;
 	memset(&in, 0 ,sizeof(in));
 	in.exp_access = access;
 	in.addr = addr;
 	in.length = length;
 	in.pd = m_p_ibv_pd;
-	ibv_mr *mr = ibv_exp_reg_mr(&in);
+	mr = ibv_exp_reg_mr(&in);
 #else
-	ibv_mr *mr = ibv_reg_mr(m_p_ibv_pd, addr, length, access);
+	mr = ibv_reg_mr(m_p_ibv_pd, addr, length, access);
 #endif
 	VALGRIND_MAKE_MEM_DEFINED(mr, sizeof(ibv_mr));
-	return mr;
+	if (NULL == mr) {
+		ibch_logerr("failed registering a memory region "
+				"(errno=%d %m)", errno);
+	} else {
+		m_mr_map_lkey[mr->lkey] = mr;
+		lkey = mr->lkey;
+
+		ibch_logfunc("dev:%s (%p) addr=%p length=%d pd=%p",
+				get_ibname(), m_p_ibv_device, addr, length, m_p_ibv_pd);
+	}
+
+	return lkey;
 }
 
-void ib_ctx_handler::mem_dereg(ibv_mr *mr)
+void ib_ctx_handler::mem_dereg(uint32_t lkey)
 {
-	if (is_removed()) {
-		return;
+	if (!is_removed()) {
+		mr_map_lkey_t::iterator iter = m_mr_map_lkey.find(lkey);
+		if (iter != m_mr_map_lkey.end()) {
```

Comment:
consider adding debug\warn if not found?

Reply:
it might be usual case during plugout